src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/AvoidZeroLengthArrayAllocations.Fixer.cs

```diff
@@ -56,25 +58,49 @@ private static async Task<Document> ConvertToArrayEmptyAsync(Document document,
                 return document;
             }
 
-            ITypeSymbol? elementType = GetArrayElementType(nodeToFix, semanticModel, cancellationToken);
+            ITypeSymbol? elementType = GetArrayElementType(nodeToFix, semanticModel, cancellationToken, out var variableName);
             if (elementType == null)
             {
                 return document;
             }
 
             SyntaxNode arrayEmptyInvocation = GenerateArrayEmptyInvocation(generator, arrayTypeSymbol, elementType).WithTriviaFrom(nodeToFix);
-            editor.ReplaceNode(nodeToFix, arrayEmptyInvocation);
+
+            if (variableName is null)
+            {
+                editor.ReplaceNode(nodeToFix, arrayEmptyInvocation);
+            }
+            else
+            {
+                editor.ReplaceNode(nodeToFix, AddElasticMarker(generator.LocalDeclarationStatement(variableName, arrayEmptyInvocation)));
+            }
+
             return editor.GetChangedDocument();
         }
 
-        private static ITypeSymbol? GetArrayElementType(SyntaxNode arrayCreationExpression, SemanticModel semanticModel, CancellationToken cancellationToken)
+        private static ITypeSymbol? GetArrayElementType(SyntaxNode arrayCreationExpression, SemanticModel semanticModel, CancellationToken cancellationToken, out string? variableName)
         {
-            var typeInfo = semanticModel.GetTypeInfo(arrayCreationExpression, cancellationToken);
-            // When Type is null in cases like 'T[] goo = { }', use ConvertedType instead (https://github.com/dotnet/roslyn/issues/23545).
-            // When Type isn't null, do not use ConvertedType. For cases like `object[] goo = new string[0]`,
-            // we want to return the string type symbol, not the object one.
-            var arrayType = (IArrayTypeSymbol)(typeInfo.Type ?? typeInfo.ConvertedType);
-            return arrayType?.ElementType;
+            var operation = semanticModel.GetOperation(arrayCreationExpression, cancellationToken);
+
+            // If we have `string[] x = { }`, we get IArrayInitializerOperation whose parent is IArrayCreationOperation.
+            if (operation.Kind == OperationKind.ArrayInitializer)
+            {
+                operation = operation.Parent;
+            }
+
+            if (operation is IArrayCreationOperation arrayCreation)
+            {
+                variableName = null;
+                return arrayCreation.GetElementType();
+            }
+            else if (operation is IVariableDeclaratorOperation { Initializer.Value: IArrayCreationOperation arrayCreationInDeclarator, Symbol.Name: var symbolName })
+            {
+                variableName = symbolName;
+                return arrayCreationInDeclarator.GetElementType();
+            }
+
+            variableName = null;
+            return null;
```

Comment:
This is only relevant for VB, I think. But I kept it here for simplicity to avoid each language defining its own `GetArrayElementType`